### DISCUSSION TITLE
Unconstrained Functions

### DISCUSSION THREAD

**Discussion by panos-xyz** [noir-lang/noir#3390] [Q&A]
### Problem

I understand that in noir, uncertainty computation is similar to the hints function in Cairo. So, can it support functions like sqrt that are more circuit-efficient and convenient for deploying applications like Uniswap?

### Happy Case

  before: 
```
    fn sqrt_test(u32 a) -> (u32) {
        u32 result = 0;
        if (a > 3) {
            result = a;
            u32 x = a / 2 + 1;
            // assume the maximum iteration is 100
            for (u32 i = 0; i < 100; i++) {
                if (x >= result) break;
                result = x;
                x = (a / x + x) / 2;
            }
        } else if (a != 0) {
            result = 1;
        }
        return result;
    }

```
after 

```
fn sqrt_test(u32 n) -> (u32) {
        u32 b = u32_sqrt(n);
        return b;
    }
```

### Additional Context

Many smart contracts require the non-deterministic calculation function of u32_sqrt, such as Uniswap and similar platforms.

### Would you like to submit a PR for this Issue?

Maybe

**Comment by jfecher** [comment]
Yes, unconstrained functions are already implemented in Noir as a feature under the same name. See the relevant section in the documentation: https://noir-lang.org/language_concepts/unconstrained.

**Reply by AlekseyFomm** [reply]
Thx! rly!))

**Reply by YURYSERGEEVICH228** [reply]
rly!) oh my god
